R/MetaboLights.R

```diff
@@ -0,0 +1,156 @@
+#'
+#' @examples
+#'
+#' ## Get the FTP path to the data set MTBLS2
+#' mtbls_ftp_path("MTBLS2")
+#'
+#' ## Retrieve available files (and directories) for the data set MTBLS2
+#' mtbls_list_files("MTBLS2")
+#'
+#' ## Retrieve the available assay files (file names starting with "a_").
+#' afiles <- mtbls_list_files("MTBLS2", pattern = "^a_")
+#' afiles
+#'
+#' ## Read the content of one file
+#' a <- read.table(paste0(mtbls_ftp_path("MTBLS2"), afiles[1L]),
+#'     header = TRUE, sep = "\t", check.names = FALSE)
+#' head(a)
+NULL
+
+#' @rdname MetaboLights-utils
+#'
+#' @export
+mtbls_ftp_path <- function(x = character(), mustWork = TRUE) {
+    if (length(x) != 1L)
+        stop("'x' has to be a single ID.")
+    res <- paste0("ftp://ftp.ebi.ac.uk/pub/databases/metabolights/",
+                  "studies/public/", x, "/")
+    if (mustWork)
+        mtbls_list_files(x)
+    res
+}
+
+#' @importFrom curl new_handle handle_setopt curl
+#'
+#' @rdname MetaboLights-utils
+#'
+#' @export
+mtbls_list_files <- function(x = character(), pattern = NULL) {
+    cu <- new_handle()
+    handle_setopt(cu, ftp_use_epsv = TRUE, dirlistonly = TRUE)
+    tryCatch({
+        con <- curl(url = mtbls_ftp_path(x, mustWork = FALSE), "r", handle = cu)
+    }, error = function(e) {
+        stop("Failed to connect to MetaboLights. No internet connection? ",
+             "Does the data set \"", x, "\" exist?\n - ", e$message,
+             call. = FALSE)
+    })
+    fls <- readLines(con)
+    close(con)
+    if (length(pattern))
+        fls[grepl(pattern, fls)]
+    else fls
+}
+
+#' retrieves and reads the/all assay data file(s) for a given MetaboLights
+#' data set and returns it/them as a `list` of `data.frame`s. The file
+#' names of the respective assay data file(s) are reported as `names()` of
+#' the returned `list`.
+#'
+#' @param x `character(1)` with the MetaboLights ID of the data set.
+#'
+#' @importFrom utils read.table
+#'
+#' @noRd
+.mtbls_assay_list <- function(x = character()) {
+    fpath <- mtbls_ftp_path(x, mustWork = FALSE)
+    a_fls <- mtbls_list_files(x, pattern = "a_")
+    res <- lapply(a_fls, function(z) {
+        read.table(paste0(fpath, z),
+                   sep = "\t", header = TRUE,
+                   check.names = FALSE)
+    })
+    names(res) <- a_fls
+    res
+}
+
+#' Extract the MS data files from MTBLS assay tables'
+#' *Derived Spectral Data File* column(s). The function checks all present
+#' columns in the provided `data.frame` and returns the content of the first of
+#' these columns with files matching the provided `pattern`.
+#'
+#' @param x `data.frame` representing the content of an *assay* ISA file from
+#'     MetaboLights.
+#'
+#' @param pattern `character(1)` with supported data types.
+#'
+#' @return `character` with the file names matching the provided pattern.
+#'
+#' @noRd
+.mtbls_derived_data_file <- function(x, pattern = "mzML$|CDF$|mzXML$") {
+    cls <- which(colnames(x) == "Derived Spectral Data File")
```

Comment:
oh btw I have seen some studies that only had raw data files, and they put mzML there... Should we check both columns ? (i'll try to find an example)

Reply:
https://www.ebi.ac.uk/metabolights/editor/MTBLS10555/files